File: Programmers.java
import java.util.ArrayList;
import java.util.List;

// 콜라츠 수열 만들기
class Solution {
    public int[] solution(int n) {
        List<Integer> list = new ArrayList<>();
        list.add(n);
        
        while(n != 1){
            if(n % 2 == 0) { // 짝수
                n = n / 2;
            } else { // 홀수
                n = 3 * n + 1;
            }
            // n = (n % 2 == 0) ? n / 2 : 3 * n + 1; // 삼항연산자
            list.add(n);
        }
        
        int[] answer =  new int[list.size()];
        for(int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        
        return answer;
    }
}


// 배열 만들기 4
import java.util.ArrayList;
import java.util.List;
class Solution {
    public int[] solution(int[] arr) {
        List<Integer> stk = new ArrayList<>();
        for(int i = 0; i < arr.length;){
            if(stk.size() == 0){
                stk.add(arr[i]); i++;
            } else if(stk.size() != 0 && stk.get(stk.size() - 1) < arr[i]){
                stk.add(arr[i]); i++;
            } else {
                stk.remove(stk.size() - 1);
            }
        }
        return stk.stream().mapToInt(i -> i).toArray();
    }
}


// 간단한 논리 연산
class Solution {
    public boolean solution(boolean x1, boolean x2, boolean x3, boolean x4) {
        boolean answer = true;
        answer = (x1 || x2) && (x3 || x4);
        return answer;
    }
}


// 글자 이어 붙여 문자열 만들기
class Solution {
    public String solution(String my_string, int[] index_list) {
        StringBuilder answer = new StringBuilder();
        
        for(int i = 0; i < index_list.length; i++) {
            int index = index_list[i];
            answer.append(my_string.charAt(index));
        }
                
        return answer.toString();
    }
}


// 글자 이어 붙여 문자열 만들기
class Solution {
    public String solution(String my_string, int[] index_list) {
        String answer = "";
        
        for(int i = 0; i < index_list.length; i++) {
            answer += my_string.charAt(index_list[i]);
        }
        
        return answer;
    }
}


// 9로 나눈 나머지
class Solution {
    public int solution(String number) {
        int answer = 0;
        
        for(int i = 0; i < number.length(); i++) {
            answer += number.charAt(i) - '0';
        }
        
        return answer % 9;
    }
}


// 9로 나눈 나머지
class Solution {
    public int solution(String number) {
        int answer = 0;
        
        for(char i : number.toCharArray()) {
            int sum = Integer.parseInt(String.valueOf(i));
            answer += sum;
        }
        
        return (answer % 9);
    }
}


// 문자열 여러 번 뒤집기
class Solution {
    public String solution(String my_string, int[][] queries) {
        String answer = "";
        char[] charArr = my_string.toCharArray();
        
        for(int i = 0; i < queries.length; i++) {
            int start =  queries[i][0];
            int end = queries[i][1];
            
            for (int j = start; j <= (start + end)/2; j++) {
                char temp = charArr[start];
                charArr[start] = charArr[end];
                charArr[end] = temp;
                start++;
                end--;
            }
        }
        
        return new String(charArr);
    }
}


// 문자열 여러 번 뒤집기
class Solution {
    public String solution(String my_string, int[][] queries) {
        String answer = "";
        char[] charArr = my_string.toCharArray();
        
        for(int i = 0; i < queries.length; i++) {
            int start =  queries[i][0];
            int end = queries[i][1];
            
            while (start < end) {
                char temp = charArr[start];
                charArr[start] = charArr[end];
                charArr[end] = temp;
                start++;
                end--;
            }
        }
        
        return new String(charArr);
    }
}


// 문자열 여러 번 뒤집기
class Solution {
    public String solution(String my_string, int[][] queries) {
        char[] charArray = my_string.toCharArray(); // 문자열을 문자 배열로 변환
        
        for (int[] query : queries) {
            int start = query[0];
            int end = query[1];
            
            // start부터 end까지의 부분 문자열을 뒤집는 작업
            while (start < end) {
                char temp = charArray[start];
                charArray[start] = charArray[end];
                charArray[end] = temp;
                start++;
                end--;
            }
        }
        
        // 문자 배열을 다시 문자열로 변환
        return new String(charArray);
    }
}


// 배열 만들기 5
class Solution {
    public int[] solution(String[] intStrs, int k, int s, int l) {
        List<Integer> resultList = new ArrayList<>();

        for (String str : intStrs) {
            if (str.length() >= s + l) {
                String substring = str.substring(s, s + l);
                int num = Integer.parseInt(substring);
                if (num > k) {
                    resultList.add(num);
                }
            }
        }

        int[] answer = new int[resultList.size()];
        for (int i = 0; i < resultList.size(); i++) {
            answer[i] = resultList.get(i);
        }
        
        return answer;
    }
}


// 배열 만들기 5
class Solution {
    public int[] solution(String[] intStrs, int k, int s, int l) {
        List<Integer> result = new ArrayList<>();
        
        for (String str : intStrs) {
            int i = Integer.parseInt(str.substring(s, s + l));
            if (i > k) {
                result.add(i);
            }
        }
        
        return result.stream().mapToInt(i->i).toArray();
    }
}


// 배열 만들기 5
class Solution {
    public int[] solution(String[] intStrs, int k, int s, int l) {
        List<Integer> list = new ArrayList<>();

        for(String str : intStrs) {
            int num = Integer.parseInt(str.substring(s, s + l));
            if(num > k) {
                list.add(num);
            }
        }
        
        int[] answer = new int[list.size()];
        
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        
        return answer;
    }
}


// 배열 만들기 5
class Solution {
    public int[] solution(String[] intStrs, int k, int s, int l) {
        return Arrays.stream(intStrs).mapToInt(value -> Integer.parseInt(value.substring(s, s + l))).filter(value -> value > k).toArray();
    }
}

File: Programmers.js
// 문자열로 변환
// 정수 n이 주어질 때, n을 문자열로 변환하여 return하도록 solution 함수를 완성해주세요.
function solution(n) {
    var str = n;
    var answer = str + '';
    return answer;
}

const solution = String

// 공배수
// 정수 number와 n, m이 주어집니다. number가 n의 배수이면서 m의 배수이면 1을 아니라면 0을 return하도록 solution 함수를 완성해주세요.
function solution(number, n, m) {
    var answer = 0;
    
    if(number % n === 0 && number % m === 0){
        answer = 1;  
    } else {
        answer = 0;
    }
    
    return answer;
}

function solution(number, n, m) {
  return +!(number % n || number % m);
}

// 첫 번째로 나오는 음수
// 정수 리스트 num_list가 주어질 때, 첫 번째로 나오는 음수의 인덱스를 return하도록 solution 함수를 완성해주세요. 음수가 없다면 -1을 return합니다.
function solution(num_list) {
       
    for(i = 0; i < num_list.length; i++){
        if(num_list[i] < 0){
            return i;
        }
    }
    
    return -1;
}

function solution(num_list) {
    var answer = 0;
    
    for(i = 0; i < num_list.length; i++){
        if(num_list[i] < 0){
            answer = i;
            break;
        }else {
            answer = -1;
        }
    }
    
    return answer;
}

const solution = num_list => num_list.findIndex(v => v < 0)

function solution(num_list) {
    var answer = num_list.findIndex(x => x < 0);
    return answer;
}

function solution(num_list) {
    return num_list.findIndex((num) => num < 0);
}

// 문자열의 앞의 n글자
// 문자열 my_string과 정수 n이 매개변수로 주어질 때, my_string의 앞의 n글자로 이루어진 문자열을 return 하는 solution 함수를 작성해 주세요.
function solution(my_string, n) {
    return my_string.substr(0, n);
}

function solution(my_string, n) {
    return my_string.slice(0, n)
 }
 
// n보다 커질 때까지 더하기
// 정수 배열 numbers와 정수 n이 매개변수로 주어집니다. numbers의 원소를 앞에서부터 하나씩 더하다가 그 합이 n보다 커지는 순간 이때까지 더했던 원소들의 합을 return 하는 solution 함수를 작성해 주세요.
function solution(numbers, n) {
    var answer = 0;
    
    for(i = 0; i < numbers.length; i++){
        if(answer <= n){
            answer += numbers[i];
        }
    }
    
    return answer;
}

function solution(numbers, n) {
    return numbers.reduce((a,c,i,t)=>(a<=n)?a+c:a)
}

function solution(numbers, n) {
    let sum = 0;
  
      for(let i = 0; i<numbers.length; i++) {
        sum += numbers[i]
        if(sum > n) break;
      }
      return sum;
  }

// 5명씩
// 최대 5명씩 탑승가능한 놀이기구를 타기 위해 줄을 서있는 사람들의 이름이 담긴 문자열 리스트 names가 주어질 때, 앞에서 부터 5명씩 묶은 그룹의 가장 앞에 서있는 사람들의 이름을 담은 리스트를 return하도록 solution 함수를 완성해주세요. 마지막 그룹이 5명이 되지 않더라도 가장 앞에 있는 사람의 이름을 포함합니다.
function solution(names) {
    return names.filter((_, i) => i % 5 === 0);
}

function solution(names) {
    var answer = [];
    
    for(i = 0; i < names.length; i += 5){
        answer.push(names[i])
    }
    
    return answer;
}

function solution(names) {
    var answer = [];
    
    for(i = 0; i < names.length; i++){
        if(i % 5 === 0){
            answer.push(names[i])    
        }
    }
    
    return answer;
}

// 정수 찾기
// 정수 리스트 num_list와 찾으려는 정수 n이 주어질 때, num_list안에 n이 있으면 1을 없으면 0을 return하도록 solution 함수를 완성해주세요.
function solution(num_list, n) {
    var answer = 0;
    
    for(i = 0; i < num_list.length; i++) {
        if(num_list[i] === n) {
            answer = 1;
        }
    }
    
    return answer;
}

function solution(num_list, n) {
  return num_list.includes(n) ? 1 : 0;
}